File: bin/anagram.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
)

type CharCount [26]byte

func getCanonicalForm(word string) string {
	runes := []rune(strings.ToLower(word))
	sort.Slice(runes, func(i, j int) bool { return runes[i] < runes[j] })
	return string(runes)
}

func stringToCharCount(s string) CharCount {
	var counts CharCount
	for _, r := range strings.ToLower(s) {
		if r >= 'a' && r <= 'z' {
			counts[r-'a']++
		}
	}
	return counts
}

func (c CharCount) canSubtract(other CharCount) bool {
	for i := 0; i < 26; i++ {
		if other[i] > c[i] {
			return false
		}
	}
	return true
}

func (c CharCount) subtract(other CharCount) CharCount {
	var result CharCount
	for i := 0; i < 26; i++ {
		result[i] = c[i] - other[i]
	}
	return result
}

func (c CharCount) isEmpty() bool {
	for i := 0; i < 26; i++ {
		if c[i] != 0 {
			return false
		}
	}
	return true
}

type WordMap struct {
	data            map[string][]string
	canonicalCounts map[string]CharCount
}

func buildWordMap(dictionaryPath string) (*WordMap, error) {
	file, err := os.Open(dictionaryPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	wordMap := make(map[string][]string)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		word := strings.TrimSpace(strings.ToLower(scanner.Text()))
		if word == "" {
			continue
		}

		// Check if word contains only letters
		isAlpha := true
		for _, r := range word {
			if r < 'a' || r > 'z' {
				isAlpha = false
				break
			}
		}

		if isAlpha {
			canonical := getCanonicalForm(word)
			wordMap[canonical] = append(wordMap[canonical], word)
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	// Precompute canonical character counts
	canonicalCounts := make(map[string]CharCount, len(wordMap))
	for canonical := range wordMap {
		canonicalCounts[canonical] = stringToCharCount(canonical)
	}

	return &WordMap{
		data:            wordMap,
		canonicalCounts: canonicalCounts,
	}, nil
}

func cleanPhrase(phrase string) string {
	var sb strings.Builder
	for _, r := range strings.ToLower(phrase) {
		if r >= 'a' && r <= 'z' {
			sb.WriteRune(r)
		}
	}
	return sb.String()
}

func findAnagrams(phrase string, wordMap *WordMap, maxWords int) {
	cleanedPhrase := cleanPhrase(phrase)
	if cleanedPhrase == "" {
		fmt.Println("Input phrase is empty or contains no letters.")
		return
	}

	initialCounts := stringToCharCount(cleanedPhrase)

	// Use a map for deduplication with mutex for concurrent-safe access
	found := make(map[string]bool)
	var mu sync.Mutex

	// Pre-filter word map to only include words that could possibly fit
	validCanonicals := make([]string, 0, len(wordMap.data))
	for canonical := range wordMap.data {
		if initialCounts.canSubtract(wordMap.canonicalCounts[canonical]) {
			validCanonicals = append(validCanonicals, canonical)
		}
	}

	// Sort canonicals by length (longer first) for better pruning
	sort.Slice(validCanonicals, func(i, j int) bool {
		return len(validCanonicals[i]) > len(validCanonicals[j])
	})

	var search func(remaining CharCount, currentWords []string)
	search = func(remaining CharCount, currentWords []string) {
		// Base case: no letters left
		if remaining.isEmpty() {
			anagram := strings.Join(currentWords, " ")
			mu.Lock()
			if !found[anagram] {
				found[anagram] = true
				fmt.Println(anagram)
			}
			mu.Unlock()
			return
		}

		// Too many words
		if len(currentWords) >= maxWords {
			return
		}

		// Try every word whose canonical letters fit inside remaining
		for _, canonical := range validCanonicals {
			countsNeeded := wordMap.canonicalCounts[canonical]

			// Skip if we can't subtract this set of letters
			if !remaining.canSubtract(countsNeeded) {
				continue
			}

			// Valid → try using each actual word
			for _, word := range wordMap.data[canonical] {
				newRemaining := remaining.subtract(stringToCharCount(word))
				newWords := make([]string, len(currentWords), len(currentWords)+1)
				copy(newWords, currentWords)
				newWords = append(newWords, word)
				search(newRemaining, newWords)
			}
		}
	}

	search(initialCounts, []string{})
}

func main() {
	maxWords := flag.Int("n", 3, "Maximum number of words allowed in the anagram")
	dictionaryPath := flag.String("dictionary", "/usr/share/dict/words", "Path to the dictionary file")
	flag.Parse()

	if flag.NArg() < 1 {
		fmt.Println("Usage: anagram [options] <phrase>")
		flag.PrintDefaults()
		os.Exit(1)
	}

	phrase := flag.Arg(0)

	wordMap, err := buildWordMap(*dictionaryPath)
	if err != nil {
		fmt.Printf("Error: Dictionary file not found or could not be read: %v\n", err)
		return
	}

	if *maxWords < 1 {
		*maxWords = 1
	}

	findAnagrams(phrase, wordMap, *maxWords)
}

File: bin/anagram.py
#!/usr/bin/python3
import argparse
from collections import Counter, defaultdict
from typing import Dict, List


def get_canonical_form(word: str) -> str:
    return "".join(sorted(word.lower()))


def build_word_map(dictionary_path: str) -> Dict[str, List[str]]:
    word_map = defaultdict(list)
    try:
        with open(dictionary_path, 'r', encoding='utf-8', errors='ignore') as f:
            for line in f:
                word = line.strip().lower()
                if word and word.isalpha():
                    canonical = get_canonical_form(word)
                    word_map[canonical].append(word)
        return dict(word_map)
    except FileNotFoundError:
        print(f"Error: Dictionary file not found at {dictionary_path}")
        return {}


def find_anagrams_generalized(phrase: str, word_map: Dict[str, List[str]], max_words: int):
    cleaned_phrase = "".join(filter(str.isalpha, phrase)).lower()
    initial_counts = Counter(cleaned_phrase)

    if not initial_counts:
        print("Input phrase is empty or contains no letters.")
        return

    # Precompute canonical Counter objects
    canonical_counters = {canonical: Counter(canonical) for canonical in word_map}

    found = set()

    def search(remaining_counts: Counter, current_words: List[str]):
        # Base case: no letters left
        if not remaining_counts:
            anagram = " ".join(current_words)
            if anagram not in found:
                found.add(anagram)
                print(anagram)
            return

        # Too many words
        if len(current_words) >= max_words:
            return

        # Try every word whose canonical letters fit inside remaining_counts
        for canonical, words in word_map.items():
            counts_needed = canonical_counters[canonical]

            # Skip if we can't subtract this set of letters
            if any(counts_needed[c] > remaining_counts[c] for c in counts_needed):
                continue

            # Valid → try using each actual word
            for word in words:
                new_remaining = remaining_counts - Counter(word)
                search(new_remaining, current_words + [word])

    search(initial_counts, [])


def main():
    parser = argparse.ArgumentParser(
        description="Finds anagrams for an input phrase using a dictionary file, supporting N-word anagrams."
    )
    parser.add_argument('-n', '--words', type=int, default=3, help="Maximum number of words allowed in the anagram")
    parser.add_argument('--dictionary', type=str, default='/usr/share/dict/words', help="Path to the dictionary file.")

    parser.add_argument('phrase', type=str, help="The input phrase to find anagrams for (e.g., 'elvis')")
    args = parser.parse_args()

    word_map = build_word_map(args.dictionary)
    if not word_map:
        return

    max_words = max(1, args.words)
    find_anagrams_generalized(args.phrase, word_map, max_words)


if __name__ == '__main__':
    main()
